Feature/add pseudo random access view

rrahn opened the pull request:
Fixes #720 

This PR adds the pseudo_random_access view and also changes the model of the pseudo random access behavior of the range. Thus, the begin_ra/end_ra interface and the second iterator definiton of the gap_decorator is not needed anymore.

eseiler commented on test/snippet/range/views/pseudo_random_access.cpp:
```diff
@@ -0,0 +1,38 @@
+    auto seq = "ACGTACGACT"_dna4;
+    seqan3::gap_decorator aligned_seq{seq};
+
+    // It does not fulfill random access semantics because it does not allow constant time access to aribtrary
```
```suggestion
    // It does not fulfil random access semantics because it does not allow constant time access to arbitrary
```

rrahn commented on include/seqan3/range/decorator/gap_decorator.hpp:
```diff
@@ -83,418 +84,14 @@ template <std::ranges::viewable_range inner_type>
-        }
-        //!\}
-    };
+    // Declaration of class's iterator types; for the defintion see below.
```
No. That should be fine.

marehr commented on include/seqan3/range/concept.hpp:
```diff
@@ -61,4 +61,68 @@ SEQAN3_CONCEPT forwarding_range = std::ranges::range<type> && requires (type &&
+ */
+//!\cond
+template <typename rng_t>
+SEQAN3_CONCEPT pseudo_random_access_range =
```
Some random thought with the naming.

`rng | views::pseudo_random_access` makes the iterator of the range (pseudo) random-accessible 
`pseudo_random_access_range` checks whether the iterator of the range is pseudo random-accessible 
`pseudo_random_access_iterator` checks whether the iterator is pseudo random-accessible

rrahn replied:
I am not sure I could follow what your suggestion is here?
But how about calling the first: `views::enforce_random_access`. I think it more clearly states the intent of the view. So pseudo_random_access_range is defined by means of the pseudo_random_access_iterator and the enforce_random_access requires either a random_access_range or a pseudo_random_access_range. Is that more appropriate?

rrahn commented on include/seqan3/range/detail/inherited_iterator_base.hpp:
```diff
@@ -171,137 +171,154 @@ class inherited_iterator_base : public std::conditional_t<std::is_pointer_v<base
+    //!\endcond
+    constexpr auto operator++(int) noexcept(noexcept(std::declval<base_t &>()++))
+    //!\cond
+        requires requires (base_t_ i) { i++; requires !std::same_as<decltype(i++), base_t_>; }
```
But what is `i ` in this context now?

marehr replied:
sry; didn't think that through

rrahn commented on include/seqan3/range/detail/inherited_iterator_base.hpp:
```diff
@@ -171,137 +171,154 @@ class inherited_iterator_base : public std::conditional_t<std::is_pointer_v<base
     constexpr derived_t & operator++() noexcept(noexcept(++std::declval<base_t &>()))
     //!\cond
-        requires std::input_iterator<base_t>
+        requires requires (base_t_ i) { ++i; }
```
Well, it certainly is unlikely that there is an iterator that does not have implemented the ++iterator, but in case it does your code will fail with a hard error instead of rejecting the function because of an unsatisfied constrained. That is because base_t is not a template argument of the function and thus it fails to compile if it does not support the ++ interface. The other solution would be to define for every operator a specific concept. But I really prefer the current solution a lot more.

rrahn commented on include/seqan3/range/views/pseudo_random_access.hpp:
```diff
@@ -0,0 +1,393 @@
+    //!\cond
+     requires !std::same_as<remove_cvref_t<viewable_rng_t>, view_pseudo_random_access> &&
+              std::ranges::viewable_range<viewable_rng_t> &&
+              std::constructible_from<urng_t, ranges::ref_view<std::remove_reference_t<viewable_rng_t>>>
```
yes, because otherwise you ran into compiler errors due to recursive compiler instantiations inside of the range adaptor definitions.

rrahn commented on include/seqan3/range/views/pseudo_random_access.hpp:
```diff
@@ -0,0 +1,393 @@
+ *
+ * \details
+ *
+ * **Header**
```
sure

rrahn commented on include/seqan3/range/views/pseudo_random_access.hpp:
```diff
@@ -0,0 +1,393 @@
+     * \{
+     */
+    // Import operator from base.
+    using base_t::operator-;
```
If you overload one of them (which we do for the distance) all other will not be visible. So you need to import them to make them available. That's at least the case for class templates.

rrahn commented on include/seqan3/range/decorator/gap_decorator.hpp:
```diff
@@ -1038,6 +578,316 @@ template <std::ranges::view urng_t>
+    //!\brief Dereference operator returns a copy of the element currently pointed at.
+    constexpr reference operator*() const noexcept
+    {
+        return (is_at_gap) ? static_cast<reference>(gap{})
```
Well, I try to change this, but note, that sometimes static_cast is required when narrowing type conversion would happen. Though, I believe this is not the case here.

rrahn commented:
@smehringer I removed you from the reviewer list as @marehr came to it first. Since much of this was already discussed in the other PR #1237 and I just moved it into another PR I thought it might not be necessary anymore to have a third review on this. But if you insist then I welcome any other review. But I'd love to get it done asap as all of this is part of the larger changes regarding the alignment refactoring which is till open.
